### ISSUE TITLE
False positive for testImplementation dependency with classifier

### ISSUE THREAD

**Issue by hakjac** [autonomousapps/dependency-analysis-android-gradle-plugin#841]
**Build scan link**
N/A

**Plugin version**
1.18.0

**Gradle version**
7.6

**(Optional) Android Gradle Plugin (AGP) version**
N/A

**`reason` output for bugs relating to incorrect advice**
```
> Task :E84HSP-app:reason

----------------------------------------
You asked about the dependency 'se.micronic.bennu:kernel:6.2.0.0-46-16c1002'.
You have been advised to remove this dependency from 'testImplementation'.
----------------------------------------

Shortest path from :E84HSP-app to se.micronic.bennu:kernel:6.2.0.0-46-16c1002 for compileClasspath:
:E84HSP-app
\--- se.micronic.bennu:kernel:6.2.0.0-46-16c1002

Shortest path from :E84HSP-app to se.micronic.bennu:kernel:6.2.0.0-46-16c1002 for runtimeClasspath:
:E84HSP-app
\--- se.micronic.bennu:kernel:6.2.0.0-46-16c1002

Shortest path from :E84HSP-app to se.micronic.bennu:kernel:6.2.0.0-46-16c1002 for testCompileClasspath:
:E84HSP-app
\--- se.micronic.bennu:kernel:6.2.0.0-46-16c1002

Shortest path from :E84HSP-app to se.micronic.bennu:kernel:6.2.0.0-46-16c1002 for testRuntimeClasspath:
:E84HSP-app
\--- se.micronic.bennu:kernel:6.2.0.0-46-16c1002

Source: main
------------
* Exposes 27 classes, 5 of which are shown: se.micronic.bennu.kernel.gui.DisplayConditionsIfc, se.micronic.bennu.kernel.logevent.LogEventDeclaration, se.micronic.bennu.kernel.logevent.LogEventId, se.micronic.bennu.kernel.logevent.XMLLogEventSpecificationBase, se.micronic.bennu.kernel.message.Command (implies api).
* Imports 3 constants: se.micronic.bennu.kernel.message.*, se.micronic.bennu.kernel.message.CommandDeclaration.*, se.micronic.bennu.kernel.variable.* (implies implementation).

Source: test
\------------
\* Uses 1 class: se.micronic.bennu.kernel.message.threaded.ThreadedCommandTestExtension (implies testImplementation).

```
**Describe the bug**
I have this dependency, (please note the classifier = 'test'), without it my project do not compile.
```
  testImplementation("se.micronic.bennu:kernel:6.2.0.0-46-16c1002") { artifact { classifier = 'test' } }
```

So, the test code of my project depends on the test code of another project which is selected using the 'classifier = test', however the dependency analysis seems to ignore the classifier.

**To Reproduce**
N/A

**Expected behavior**
Should not give advise to remove the dependency since it is used for compilation of test code.

**Additional context**
N/A

**Comment by autonomousapps** [comment]
Thanks for the issue. I think it relates to https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/issues/342.

@jjohannes what do you think? Is this something that would fall under the umbrella of work you'd like to do?

**Comment by jjohannes** [comment]
Yes this and #342 are related to the "capability" topic (Case 2 here https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/issues/298#issuecomment-1346159435). Using a classifier is like addressing another feature variant (only that the feature variant is not described in the corresponding POM metadata). IIRC Gradle does some tricks to still represent a Jar with classifier as a feature variants internally.

If I start looking into #298 and related issues, I can take a look if these issues can be solved together.

**Comment by jjohannes** [comment]
In the context of #854 I added a test to reproduce this.

The problem that can cause _false positives_ is something I missed in #688.

The situation is that one component can consist of multiple Jar files. The code was initially assuming that there is always only one Jar file per component (or, to be more precise, variant of a component). With #688, the fact that there may be more than one Jar (in a non-standard setup) is respected. But there is a bug. The "capabilities" are computed for each Jar, but then the computation are merged later and the "capability details" (like which Public Methods exist) of one of the Jars is dropped. The fix is part of #854 (see https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/pull/854#discussion_r1105859793).

Now in the case of classifiers, as pointed out in the comments of #342, Gradle looses the classifier information, which is partly a conceptual problem of how this whole classifier thing (invented by Maven) works. It's better to use full-fledged "feature variants" where you can which the plugin will support with #854.

But if classifiers are at play, for the plugin it now looks as if all Jars (e.g. the `kernel.jar` and `kernel-test.jar` ) belong directly to the component. And they are treated as if they were one Jar. With the fix, this should work. Then there should be no more false positives (that lead to code that won't compile anymore).

Still advices involving classifiers can be wrong. But I would say this is something to follow on in #342.
